staging/src/k8s.io/api/storage/v1beta1/types.go

```diff
@@ -209,3 +209,203 @@ type VolumeError struct {
+// volume driver deployed on the cluster.
+// CSI drivers do not need to create the CSIDriver object directly. Instead they may use the
+// cluster-driver-registrar sidecar container. When deployed with a CSI driver it automatically
+// creates a CSIDriver object representing the driver.
+// Kubernetes attach detach controller uses this object to determine whether attach is required.
+// Kubelet uses this object to determine whether pod information needs to be passed on mount.
+// CSIDriver objects are non-namespaced.
+type CSIDriver struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object metadata.
+	// metadata.Name indicates the name of the CSI driver that this object
+	// refers to; it MUST be the same name returned by the CSI GetPluginName()
+	// call for that driver.
+	// The driver name must be 63 characters or less, beginning and ending with
+	// an alphanumeric character ([a-z0-9A-Z]) with dashes (-), dots (.), and
+	// alphanumerics between.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the CSI Driver.
+	Spec CSIDriverSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// items is the list of CSIDriver
+	Items []CSIDriver `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+	// attachRequired indicates this CSI volume driver requires an attach
+	// operation (because it implements the CSI ControllerPublishVolume()
+	// method), and that the Kubernetes attach detach controller should call
+	// the attach volume interface which checks the volumeattachment status
+	// and waits until the volume is attached before proceeding to mounting.
+	// The CSI external-attacher coordinates with CSI volume driver and updates
+	// the volumeattachment status when the attach operation is complete.
+	// If the CSIDriverRegistry feature gate is enabled and the value is
+	// specified to false, the attach operation will be skipped.
+	// Otherwise the attach operation will be called.
+	// +optional
+	AttachRequired *bool `json:"attachRequired,omitempty" protobuf:"varint,1,opt,name=attachRequired"`
+
+	// If specified, podInfoRequiredOnMount indicates this CSI volume driver
+	// requires additional pod information (like podName, podUID, etc.) during
+	// mount operations.
+	// If value is not specified, pod information will not be passed on mount.
```

Comment:
My real question didn't get answered: Why is this an option? Why would we NOT pass this in all cases?

Reply:
Drivers may validate that all the options passed in are valid. So it might be a breaking change to some drivers if we start passing them attributes we didn't ask for. I do agree though that the API would be cleaner if we didn't need to specify it at all. The attributes are prefixed, so we could say that all drivers should handle or ignore attributes with that prefix. I'm not sure how we can do that without breaking backwards compatibility though.

Reply:
I see.  That's unfortunate.  Do  we know if it is really an issue for existing drivers, or just theoretical?

Can we add a sentence here explaining this?

Who sets this field - does the driver itself announce this or does a human need to link it together?

why is this versioned rather than a bool?  Do we actually have plans to introduce other versions?

Reply:
There are definitely some drivers out that do this [like this silly one](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/gce-pd-csi-driver/controller.go#L113).

The csi spec was vague on what a driver should do when given an invalid parameter, so some do strict validation, and some ignore it.
```
The Plugin is responsible for parsing and validating these parameters
```

The driver specifies the option as part of driver deployment; it's not a capability in the csi spec though, it's an option to the driver sidecar.

Originally, the thought was if we want to add more metadata in the future, then it would be bumped to a new version.  However, I think because we prefix all these attributes names now, we could make it a bool and just say "everything we give you will have this prefix".  That way drivers don't need to explicitly bump versions whenever we make an update.

That also assumes that we will always add fields, and never remove.

Reply:
The packager of the driver sets the filed indicating it is safe (or even required) to pass this data through.

It was versioned because someone thought it might be possible that an incompatible set of attributes could be passed some day. So it the driver packager states which set of attributes should be passed.

Reply:
During the review of the initial alpha implementation, it was [suggested to use a versioned string](https://github.com/kubernetes/kubernetes/pull/67803#discussion_r214470294) instead of a bool.

That being said, I don't see much value here for making it a versioned string unless:
* we want to pass completely different sets of attributes in certain scenarios
* we want to delete/deprecate attributes

I don't really see a need for this now, so I am fine with changing it back to a bool.

Reply:
If we change this back to a bool, we can't differentiate the attributes based on versions any more?  Should we keep it as a version string so we still have that flexibility?

Reply:
I don't know what purpose differentiating solves.  If we ever need to version this we can add fields, but we can't take them away.

Less is more?

Reply:
Sure.  Will change to bool.